GPMSA functional branch

roystgnr opened the pull request:
This isn't ready to merge yet, because it's still missing big chunks of implementation, but it should be ready to look at, so I don't hit @dmcdougall with one giant pile of commits out of the blue later.  I'll keep pushing further commits here.

In particular, the APIs are there now for specifying a mesh for functional data, for supplying point locations for functional data in experimental outputs, and for mixing multiple functional data fields with multivariate data.

roystgnr committed:
Start "numExperimentOutputs != constant" support

This won't be working until the interpolation code is finished, but it
at least strips out a lot of bad assumptions without changing any
regression test solutions.

roystgnr committed:
Add GPMSA classes for meshing, interpolation

The interpolation code isn't finished yet, but these should be at a
state where they compile.

roystgnr committed:
Build new GPMSA mesh code

roystgnr committed:
Add TensorProductMesh interpolation code

I tried to make this as general as possible.  The benefits of this
design are that 1-D problems aren't slowed down by 4-D interpolation
code and that 5-D problems will be trivial to support if we ever need.

The cost of this design is that it's much harder to understand than a
couple quadruply-nested loops would have been.  I should probably add
some cppunit tests...

roystgnr committed:
Move first_solution_index to base class, size_t

We'll need to query this later when determining what meshes to use to
interpret what data, in cases of mixed multivariate+functional or
functional+other-functional data.  We might some day want to support
more than 2^32 function outputs, too; let's avoid adding more code
which assumes otherwise.

roystgnr committed:
Fixup iterator search

std::lower_bound is deceivingly named, but I think this should be
correct now.

We still shouldn't trust this code until I get some unit tests in.

roystgnr committed:
Add SimulationMesh index information getters

roystgnr committed:
GPMSAFactory::addSimulationMesh()

We don't actually use it yet, but we're getting closer!

It could be easy for users to not meet our requirements, and it's
cheap to check, so let's check even when not debugging.

roystgnr committed:
Pass simulation mesh(s) to GPMSAEmulator

roystgnr committed:
Track data point locations for GPMSA functional

roystgnr committed:
Supply experimentVariables indices to GPMSA

This lets us have multiple functional outputs, or multivariate
simulations with "sparse" experimental results, or mixtures of
functional and multivariate data.

roystgnr committed:
Clarify addSimulationMesh, indexing requirements

roystgnr committed:
Const SimulationOutputMesh methods should be const

roystgnr committed:
Handle autoscaling for functional, mixed GPMSA

roystgnr commented:
Oh, and if anyone actually pulls this rather than just examining it via the web interface, be ready for lots of rebase+force-push to deal with.  So far I've had "rebase to sync with Damon's fixes", "rebase to fixup a dbg-mode bug since I'd forgotten to test in dbg mode", and I may shortly have a "rebase to fix a bug that seems to have cropped up when I didn't correctly fix conflicts with Damon's fixes".

roystgnr committed:
Avoid is_sorted, since it's C++11 only

We'll be moving to C++11 eventually, but right now our Travis builds
are C++98.

roystgnr committed:
Explanation, options vars for gaussian kernels

roystgnr committed:
Fix asserts on C++98

Apparently C++11 will let me output iterators but 98 won't.

roystgnr committed:
Fix and parse gaussian discrepancy kernel params

roystgnr committed:
Add+parse gaussian discrepancy support threshold

This will be needed for auto-generation of gaussian discrepancy
kernels.

roystgnr committed:
SimulationOutputMesh generateDiscrepancyBases shim

We ought to be able to do these independent of each other and of
multivariate discrepancies, but we do need to do these in the mesh
code if we want to be able to do it on non-tensor-product meshes
eventually.

roystgnr committed:
Use continue to reduce loop nesting

This makes the code a little more readable, I think.

roystgnr committed:
TensorProductMesh::generateDiscrepancyBases()

Now implemented, at least in the non-periodic case.

roystgnr committed:
Normalize gaussian discrepancy bases

I kind of think normalizing with peak value = 1 would be better than
normalizing by l2 norm = 1, but I *definitely* think we should try to
match the Matlab code for our initial tests.

roystgnr committed:
Normalize based on transpose indexing

Okay, *this* should be what matches the Matlab code.

roystgnr committed:
Small discrepancy matrices can be temporaries

We don't use them directly later, we use the matrix we built from
them.

roystgnr committed:
Remove now-unused discrepancyMatrices member

We replaced this with local temporaries so we don't need it anymore.

roystgnr commented:
We have got to turn those codecov red X's into something else, in the icons next to commit hashes.  I know admitting I have a problem is the first step, and I admit I have a problem with test coverage, but surely there's a huge difference between "PR is bad because it breaks tests" (which should get a red X) and "PR is bad because it adds features before adding test coverage" (which, I dunno, frowny face?)

roystgnr committed:
Set m_experimentPoints/Variables when provided

roystgnr committed:
Interpolate experiment data on discrepancy bases

roystgnr commented:
Passed "make check" on both my systems and Travis' after the first rebase; let's see if Travis likes the second rebase too.

roystgnr committed:
Interpolate K matrix data in functional case

This should be refactored a bit, but in the meantime I'm just happy it
doesn't regress any of the multivariate tests

roystgnr committed:
Discrepancy groups in functional/mixed cases

We no longer assume *either* 1 discrepancy group or numOutputs
discrepancy groups.

We still assume 1 discrepancy group per mesh-or-multivariate-scalar,
but that's probably good enough for a while.

roystgnr committed:
Functional data has a single, not per-point, mean

roystgnr commented:
This is now at the point where it's not impossible that a functional case might work.

(In practice, I'm sure the first functional case I try will segfault if optimized or throw an exception if GLIBCXX_DEBUG; we'll find out for sure and hopefully fix what ought to be that easiest-to-hunt-down class of errors soon)

roystgnr committed:
Remove GPMSA experimentSpace

We no longer need it, and it no longer makes sense, since each
experiment can live in a different output space now.

roystgnr commented:
Ah, there's still a clear bug here - I'm calling output normalization functions with vector indices rather than variable numbers.  I'll fix that.

roystgnr committed:
Clarify scaling method indexing

In functional and mixed cases we want ask the user to set scaling
options per-variable, where each variable may encompass more than one
output index, and so the vector indexing won't match the output
indexing.

roystgnr committed:
Fix scaling in functional/mixed cases

We leave the scaling methods using output vector indexing, but since
we're storing scaling data based on variable indexing we need to
translate from one to the other.  We'll cache the translation for
speed.

roystgnr commented:
Ah, hell, that was another merge conflict with Damon's fixes.  Rebasing and manually fixing now.

roystgnr commented:
And there's another merge conflict.  We definitely ought to just get this into dev ASAP.

roystgnr commented:
Okay, two fixups to previous commits, and it's compiling and passing regression tests for me again.

I'm still hoping to merge this as soon as we get the Wy fix in, but it can wait if @dmcdougall and others want more time to review.  At the rate I'm rebasing I'm soon to dethrone Harry Turtledove as the master of alternate history.

roystgnr committed:
Clear up comment grammar

roystgnr committed:
Zero out new simulation points

This is just safer, and the slowdown should be trivial.

roystgnr commented:
This isn't actually quite ready yet - it doesn't even compile (well, link) with --enable-boost-program-options, apparently because I added a couple new use cases without adding corresponding template specializations.

roystgnr committed:
Fill GPMSA gaussian discrepancy options

If we don't get these options from the input file, don't leave the
user with truncated vectors, fill the vectors with defaults.

We can't do this ahead of time, because we don't know for certain
until the last minute how many simulation meshes the user is going to
attach.

roystgnr committed:
Fix empty discrepancy option vector printing

If we don't attach any simulation meshes for functional data, or if we
print GPMSAOptions before the mesh-related options are finalized, then
we have empty vectors here, and we need to print the closing brace
anyway.

roystgnr committed:
Move a GPMSA assert

If we're getting bad results here it's almost certainly because of a
bad covariance matrix, so we should let the code get to the point
where that gets printed before we potentially fail an assertion and
quietly die.

roystgnr committed:
test_gpmsa_functional regression test

Now we can make sure we're getting the same wrong answers on every
system.

roystgnr committed:
Add GPMSA functional "gold standard" results

These will surely change as we debug, but at least we can now make
sure we don't change them by accident.

roystgnr committed:
Run GPMSA functional regression test

roystgnr commented:
We have a regression test for the functional case now.

Well, we have a progression test, anyway, which will become a regression test once we've actually worked all the bugs out.

dmcdougall commented:
Do you want me to wait for you to fix the conflict before I start giving feedback?

roystgnr commented:
Not just that conflict, but also let's get the Wy fix in, and then that ought to be the last rebase I need to do and we'll actually have some stable commits to attach comments to.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -160,7 +160,11 @@ GPMSAEmulator<V, M>::lnValue(const V & domainVector,
                         (this->num_svd_terms < num_nonzero_eigenvalues) +
+=======
+                        (this->num_svd_terms < numSimulationOutputs) +
+>>>>>>> Start "numExperimentOutputs != constant" support
```
You committed some conflict markers here.

roystgnr replied:
Yeah; missed something in the rebase-before-last.  Fixed now.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -236,7 +240,11 @@ GPMSAEmulator<V, M>::lnValue(const V & domainVector,
         dimParameter + (this->num_svd_terms < num_nonzero_eigenvalues) + num_svd_terms;
+=======
+        dimParameter + (this->num_svd_terms < numSimulationOutputs) + num_svd_terms;
+>>>>>>> Start "numExperimentOutputs != constant" support
```
You committed some conflict markers here.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -280,7 +288,11 @@ GPMSAEmulator<V, M>::lnValue(const V & domainVector,
                          (this->num_svd_terms<num_nonzero_eigenvalues)];
+=======
+                         (this->num_svd_terms<numSimulationOutputs)];
+>>>>>>> Start "numExperimentOutputs != constant" support
```
You committed some conflict markers here.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -304,7 +316,11 @@ GPMSAEmulator<V, M>::lnValue(const V & domainVector,
           (this->num_svd_terms<num_nonzero_eigenvalues);
+=======
+          (this->num_svd_terms<numSimulationOutputs);
+>>>>>>> Start "numExperimentOutputs != constant" support
```
You committed some conflict markers here.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -328,7 +344,11 @@ GPMSAEmulator<V, M>::lnValue(const V & domainVector,
                                                    (num_svd_terms<num_nonzero_eigenvalues) +
+=======
+                                                   (num_svd_terms<numSimulationOutputs) +
+>>>>>>> Start "numExperimentOutputs != constant" support
```
You committed some conflict markers here.

dmcdougall commented on src/gp/inc/SimulationOutputMesh.h:
```diff
@@ -0,0 +1,56 @@
+
+  virtual void interpolateOutputs(const V & solutionVector,
+                                  const std::vector<SimulationOutputPoint> & outputPoints,
+                                  V & interpolatedValues);
```
I'll assume you add documentation for these and the other class in future commits.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -0,0 +1,122 @@
+  (unsigned int i,
+   std::vector<double> & _coord_vals)
+{
+  // We passed that argument by value, so it should be safe to steal.
```
Are you talking `_coord_vals`?  If so, that's passed by reference.

roystgnr replied:
It's passed by reference here.  It's passed by value in each of the functions that call here, so this is indeed a reference to a newly-created (and soon to expire) value.

roystgnr replied:
And this is a private helper method so we shouldn't have to worry about users misusing it and getting bitten by the stealing.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -0,0 +1,122 @@
+          indices[dim] = lb - _coordinate_vals[dim].begin();
+        }
+    }
+}
```
Note to self: this method doesn't do anything (useful) here.

dmcdougall replied:
Note to self (on a later commit): this method now does something (useful).

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -111,8 +142,51 @@ TensorProductMesh<V>::interpolateOutput
+          stride *= _coordinate_vals[raw_dim].size();
+          coefficient *= on_larger_point ?
+                         (1.0 - lb_fraction[raw_dim]) :
+                         lb_fraction[raw_dim];
```
You're right, I don't understand this code.  Tests would be nice (I assume they make a cameo later).

roystgnr replied:
Again with the reasonable-but-untrue assumptions.  I'll add comments and unit tests.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -136,14 +136,15 @@ TensorProductMesh<V>::interpolateOutput
-          std::vector<double>::const_iterator ub = lb;
-          ub++;
+          std::vector<double>::const_iterator lb = ub;
+          lb--;
```
Good grief.  When I read it the first I thought it was right (and that was after reading the documentation for `std::lower_bound`)!  Reading comprehension fail.

dmcdougall commented on src/gp/inc/SimulationOutputMesh.h:
```diff
@@ -42,6 +42,9 @@ class SimulationOutputMesh
 
+  std::size_t first_solution_index() { return _first_solution_index; }
+  virtual std::size_t n_outputs() = 0;
+
```
I assume documentation is added later.

roystgnr replied:
I can't tell whether you actually believe this, or whether you're merely trying to gently coax me into doing the right thing via lofty expectations inducing shame.

I'll tack a documentation commit or three onto the end of the PR.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -161,11 +163,7 @@ GPMSAEmulator<V, M>::lnValue(const V & domainVector,
                         (this->num_svd_terms < num_nonzero_eigenvalues) +
-=======
-                        (this->num_svd_terms < numSimulationOutputs) +
->>>>>>> Start "numExperimentOutputs != constant" support
```
Not sure what you want to do about these.

roystgnr replied:
Ah, hell, they're still in the history, aren't they?  I'll see what I can do.

roystgnr replied:
Okay, I believe I got it (as well as one conflict that didn't show up via git parsing)

dmcdougall commented on src/gp/inc/GPMSA.h:
```diff
@@ -71,7 +71,8 @@ class GPMSAEmulator : public BaseScalarFunction<V, M>
                 const M & BT_Wy_B_inv_in,
                 const M & KT_K_inv_in,
-                const GPMSAOptions & opts);
+                const GPMSAOptions & optsi,
```
Is `optsi` a turd from entering insert mode in vim?

roystgnr replied:
Of course it is.  And it's in a declaration so it compiles just fine...  :-P

dmcdougall commented on src/gp/inc/GPMSA.h:
```diff
@@ -125,10 +127,21 @@ class GPMSAEmulator : public BaseScalarFunction<V, M>
 
+  // Point(s) at which experimental data is located.
+  const std::vector<typename SharedPtr<std::vector<SimulationOutputPoint> >::Type> & m_experimentPoints;
+
```
Thanks for the comments.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -1140,14 +1145,40 @@ void
+                                 (*experimentVariables)[i].size());
+        }
+    }
+
```
Note to self: these `if` and `for` blocks don't seem to do anything (useful) here.

roystgnr replied:
Huh?  They're sanity checking consistency between the user's inputs to the API.

dmcdougall replied:
Yes that's right.  When I made that comment, the commit I was commenting on didn't do anything with the sanity-checked input.

dmcdougall commented on src/gp/inc/GPMSA.h:
```diff
@@ -277,7 +277,12 @@ class GPMSAFactory
+  // We currently assume that meshes are added in the order in which
+  // their data appears in a simulation output vector, and we assume
+  // that all multivariate data in simulation outputs comes after all
+  // functional data.
```
Good.  Thanks.

dmcdougall commented on src/gp/inc/GPMSAOptions.h:
```diff
@@ -268,6 +268,20 @@ class GPMSAOptions
+  std::vector<double> m_gaussianDiscrepancyDistanceX,
+                      m_gaussianDiscrepancyDistanceY,
+                      m_gaussianDiscrepancyDistanceZ,
+                      m_gaussianDiscrepancyDistanceT;
```
The same distance for each kernel in a single coordinate direction?

roystgnr replied:
Right; only supporting equispaced kernels for now.

roystgnr replied:
Well, equispaced or "completely 100% custom" - you can hand GPMSA your own discrepancy kernels if you want to go non-equispaced or non-tensor-product or differently-normalized or whatever.

dmcdougall commented on src/gp/src/GPMSAOptions.C:
```diff
@@ -923,8 +1071,59 @@ GPMSAOptions::print(std::ostream& os) const
+         os << ((i+1 == size) ? '}' : ',');
+       }
+
+     os << std::endl;
```
I suppose we really ought to have some convenience functions for printing vectors...

roystgnr replied:
Yeah, but so should C++98, so I'm not feeling completely awful about it.

dmcdougall commented on src/gp/inc/SimulationOutputMesh.h:
```diff
@@ -52,6 +56,9 @@ class SimulationOutputMesh
 
+  virtual void generateDiscrepancyBases(const GPMSAOptions & opt,
+                                        std::vector<typename SharedPtr<V>::Type> bases) const = 0;
+
```
Documentation is incoming later, I assume.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -140,8 +144,135 @@ template <class V>
+        }
+    }
+
   queso_not_implemented(); // FIXME
```
Not to self: this method still throws an exception.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -140,8 +144,135 @@ template <class V>
+  kernel_periodic[0] = opt.m_gaussianDiscrepancyPeriodicX[mesh_number];
+  kernel_periodic[1] = opt.m_gaussianDiscrepancyPeriodicY[mesh_number];
+  kernel_periodic[2] = opt.m_gaussianDiscrepancyPeriodicZ[mesh_number];
+  kernel_periodic[3] = opt.m_gaussianDiscrepancyPeriodicT[mesh_number];
```
I can't think of a situation where you'd want periodic kernels in one mesh but not another (assuming the same physical variable).

roystgnr replied:
One of the points of having multiple meshes is to allow for different physical variables, but it's also reasonable for different simulated experiments to discretize the same variable in different domains, including using polar coordinates for one but in rectilinear coordinates for another.

dmcdougall replied:
Ah yes.  Thanks.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -140,8 +144,135 @@ template <class V>
+            peak - _coordinate_vals[dim].front();
+          if (boundary_distance < 0)
+            {
+              const double boundary_value = std::exp(-(boundary_distance * boundary_distance)/2/stddev/stddev);
```
You're missing the normalisation constant.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -140,8 +144,135 @@ template <class V>
+            _coordinate_vals[dim].back() - peak;
+          if (boundary_distance < 0)
+            {
+              const double boundary_value = std::exp(-(boundary_distance * boundary_distance)/2/stddev/stddev);
```
You're missing the normalisation constant.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -140,8 +144,135 @@ template <class V>
+              const double peak = gaussian_peak_index[raw_dim];
+              const double point = mesh_point_index[raw_dim];
+              const double distance = peak - point;
+              gaussian_value *= std::exp(-(distance * distance)/2/stddev/stddev);
```
You're missing the normalisation constant.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -140,8 +144,135 @@ template <class V>
+    }
+
+  // Start constructing bases
+  double gaussian_peak_index[max_coordinates] = {0, 0, 0, 0};
```
This will break if we ever change `max_coordinates`.

roystgnr replied:
Not necessarily!  See http://en.cppreference.com/w/cpp/language/aggregate_initialization - if there's a list initialization, but the list is too short, then subsequent array entries are *still* initialized, to 0 for integers or floating point numbers.

On the other hand, that's not exactly common C++ knowledge, and it doesn't work if we make max_coordinates *smaller*, so I'll fix it and comment it.

dmcdougall commented on src/gp/src/TensorProductMesh.C:
```diff
@@ -273,7 +275,21 @@ TensorProductMesh<V>::generateDiscrepancyBases
+    {
+      V & basis = *bases[b];
+      basis *= invmaxnorm;
+    }
```
Hmmm, I left a couple of comments pointing out you're missing the normalisation constant.  I take it those comments are now irrelevant.

roystgnr replied:
In theory, yes.

In practice those comments may have completely vanished, because I get "Sometimes commits can disappear after a force-push." messages when trying to look at them in context...

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -1182,6 +1182,9 @@ GPMSAFactory<V, M>::addExperiments(
                              experimentVariables->size());
+
+      this->m_experimentPoints = *experimentPoints;
+      this->m_experimentVariables = *experimentVariables;
```
And you've already done sanity checking.  Good.

dmcdougall commented on src/gp/src/GPMSA.C:
```diff
@@ -1073,7 +1071,30 @@ GPMSAFactory<V, M>::setUpEmulator()
+                      singular_vector[var + variable_index_to_output_index];
+                }
+
+              B(i,j) = Kvalue;
```
And this doesn't regress anything?  That's awesome.

roystgnr replied:
None of my commits were made before tests were passed.  No idea about the intermediate stages after rebase-after-rebase, though.

dmcdougall replied:
Yeah that's fine.

roystgnr committed:
Add documentation for SimulationOutputMesh

roystgnr committed:
size-independent 0-initialization

With a comment to make it clear that's what we're doing.

roystgnr committed:
SimulationOutputPoint, TensorProductMesh docs

roystgnr committed:
More comments in TensorProductMesh calculations

roystgnr committed:
Fix "vim insert mode turd"

Thanks to @dmcdougall for finding this.  Not going to rewrite history
with it since it didn't affect compilation or output.

roystgnr commented:
Can anyone replicate the regression test failure?  It works fine for me, naturally...

roystgnr commented:
But the new unit tests I'm working on don't work, so we'll fix up those bugs first and see what happens.

roystgnr committed:
Fix iteration through TensorProductMesh dimensions

roystgnr committed:
Handle interpolation at domain min/max

roystgnr committed:
Bugfix TensorProductMesh shape function selection

roystgnr committed:
TensorProductMesh initial unit test

roystgnr committed:
Build and run TensorProductMesh unit test

roystgnr commented:
Wait a minute!  "make check" is *passing* on Travis; it's "make distcheck" that's failing.  Let's see if I can replicate *that*.  I'm probably just forgetting to install some config file or script file.

roystgnr committed:
Don't forget to EXTRA_DIST new regression gold

roystgnr commented:
OH MY GOD ALL THE CHECKS PASS MERGE IT MERGE IT MERGE IT NOW!

But seriously, @dmcdougall, I think we're good to go.  Is there anything left I should change first?
